### PR TITLE
Always create new identity for view owner

### PR DESCRIPTION
**Pull request by kokosing**

### CODE REVIEW

**Comment by kokosing** [comment]
Supersedes https://github.com/trinodb/trino/pull/13991 and https://github.com/trinodb/trino/pull/14130

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -4378,11 +4378,17 @@ private RelationType analyzeView(Query query, QualifiedObjectName name, Optional
                 // run view as view owner if set; otherwise, run as session user
                 Identity identity;
                 AccessControl viewAccessControl;
-                if (owner.isPresent() && !owner.get().getUser().equals(session.getIdentity().getUser())) {
+                if (owner.isPresent()) {
                     identity = Identity.from(owner.get())
                             .withGroups(groupProvider.getGroups(owner.get().getUser()))
                             .build();
-                    viewAccessControl = new ViewAccessControl(accessControl, session.getIdentity());
+                    if (owner.get().getUser().equals(session.getIdentity().getUser())) {
```
[the remark this answers is not in the document]
Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:4385]: Yeah, "don't restrict owner's access" is kinda vague... And this rule that view owner does not require grant option is kinda obscure. Probably a good idea to clarify.

File: core/trino-main/src/test/java/io/trino/server/TestHttpRequestSessionContextFactory.java [hunk 2]
```diff
@@ -86,7 +86,14 @@ private static void assertSessionContext(ProtocolHeaders protocolHeaders)
         assertEquals(context.getCatalog().orElse(null), "testCatalog");
```
Comment (kokosing) [inline core/trino-main/src/test/java/io/trino/server/TestHttpRequestSessionContextFactory.java:86]: Restore the authoriship

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:4385]: I agree that "Always create new identity for view owner" conveys no meaning. It's not terrible -- a person reading commit log may understand that the title doesn't hint and the change meaning and will proably read the full message. but we could do better.

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 3]
```diff
@@ -93,6 +100,12 @@ protected QueryRunner createQueryRunner()
                 .setSchema("default")
                 .build();
         DistributedQueryRunner queryRunner = DistributedQueryRunner.builder(session)
+                .setAdditionalModule(binder -> {
+                    newOptionalBinder(binder, SystemSecurityMetadata.class)
+                            .setBinding()
+                            .to(TestingSystemSecurityMetadata.class)
```
Comment (findepi) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:106]: Does this change meaning of all pre-existing tests in this class?

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 4]
```diff
@@ -299,6 +312,60 @@ public void testViewColumnAccessControl()
...
+                new TrinoPrincipal(USER, viewOwnerSession.getUser()));
+
+        assertAccessAllowed(
+                viewOwnerSession,
+                "CREATE VIEW " + columnAccessViewName + " AS SELECT * FROM orders",
+                privilege("orders", CREATE_VIEW_WITH_SELECT_COLUMNS));
+
+        // verify that role permissions are taken for the view owner's role grants
+        // even if executed by the owner with different roles enabled
+        executeExclusively(() -> {
```
Comment (findepi) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:342]: I think this class fakes parallel execution. It would run faster if it was just singlethreaded.

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 5]
```diff
@@ -299,6 +312,60 @@ public void testViewColumnAccessControl()
...
+            systemSecurityMetadata.grantRoles(getSession(), Set.of("deniedrole"), Set.of(new TrinoPrincipal(USER, "test_view_access_owner")), false, Optional.empty());
+            try {
+                getQueryRunner().getAccessControl().denyIdentityTable((identity, table) -> !identity.getEnabledRoles().contains("deniedrole") || !"orders".equals(table));
+                // TestingAccessControlManager works by denying accesses, so:
+                // * viewOwnerSession does not have deniedrole enabled
+                // * the owner has a role grant to deniedrole
+                // we deny access to deniedrole, so:
+                // * even though viewOwnerSession would be allowed,
+                // * the session created for view owner will be denied
```
Comment (findepi) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:351]: Are we testing TestingAccessControlManager? I..e. testing the testing code?

File: testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java [hunk 6]
```diff
@@ -299,6 +312,60 @@ public void testViewColumnAccessControl()
...
+                assertThatThrownBy(() -> getQueryRunner().execute(viewOwnerSession, "SELECT * FROM " + columnAccessViewName))
+                        .hasMessageMatching("Access Denied: Cannot select from columns \\[.*] in table or view \\w+\\.\\w+\\.orders");
+
+                // verify view can be queried when the role grant is revoked
+                systemSecurityMetadata.revokeRoles(getSession(), Set.of("deniedrole"), Set.of(new TrinoPrincipal(USER, "test_view_access_owner")), false, Optional.empty());
+                getQueryRunner().execute(viewOwnerSession, "SELECT * FROM " + columnAccessViewName);
+            }
+            finally {
+                getQueryRunner().getAccessControl().reset();
```
Comment (findepi) [inline testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:360]: In this class, this is typicaly done by `assertDenied` or similar, so that test methods don't need `try-finally`

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 7]
```diff
@@ -4522,7 +4522,7 @@ private void analyzeRowFilter(String currentIdentity, Table table, QualifiedObje
             ExpressionAnalysis expressionAnalysis;
             try {
                 expressionAnalysis = ExpressionAnalyzer.analyzeExpression(
-                        createViewSession(filter.getCatalog(), filter.getSchema(), Identity.forUser(filter.getIdentity()).build(), session.getPath()), // TODO: path should be included in row filter
+                        createViewSession(filter.getCatalog(), filter.getSchema(), Identity.ofUser(filter.getIdentity()), session.getPath()), // TODO: path should be included in row filter
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:4525]: 👍
 Let's maybe move it to separate PR?

File: core/trino-main/src/test/java/io/trino/server/TestHttpRequestSessionContextFactory.java (hunk 2 above)
Comment (findepi) [reply core/trino-main/src/test/java/io/trino/server/TestHttpRequestSessionContextFactory.java:86]: is this resolved?

File: core/trino-spi/src/main/java/io/trino/spi/security/Identity.java [hunk 8]
```diff
@@ -130,13 +130,17 @@ public boolean equals(Object o)
             return false;
         }
         Identity identity = (Identity) o;
-        return Objects.equals(user, identity.user);
+        return Objects.equals(user, identity.user) &&
+               Objects.equals(groups, identity.groups) &&
+               Objects.equals(principal, identity.principal) &&
+               Objects.equals(enabledRoles, identity.enabledRoles) &&
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/security/Identity.java:136]: From commit message

> The `StatementAnalyzer`, during the analysis stage, collects all table
references along with their corresponding access controls and
identities. At the end of the analysis, this is used to do the access
checks. They are collected in a `Map`, where the key is the pair of
access control and `Identity`, so that we deduplicate the necessary
checks. But for this deduplication to work properly, `Identity`
instances that result in different accesses should be considered
distinct.


is this backed by any tests? i.e. any tests that would involve StatementAnalyzer?

Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/security/Identity.java:136]: Also, this looks orthogonal to "Always create new identity for view owner" commit.
correct me if am wrong, but maybe this could be separate PR?

File: core/trino-main/src/main/java/io/trino/security/AccessControlUtil.java [hunk 9]
```diff
@@ -67,47 +59,4 @@ public static void checkCanKillQueryOwnedBy(Identity identity, Identity queryOwn
         }
         accessControl.checkCanKillQueryOwnedBy(identity, queryOwner);
     }
-
-    private static class FullIdentityEquality
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/security/AccessControlUtil.java:71]: let's squash "Remove obsolete AccessControlUtil.FullIdentityEquality" with the change to `Identity.equals`

File: core/trino-spi/src/test/java/io/trino/spi/security/TestIdentity.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.util.Optional;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/*
```
Comment (electrum) [inline core/trino-spi/src/test/java/io/trino/spi/security/TestIdentity.java:25]: Duplicate header

**Review by kokosing — COMMENTED** [review]
Addressed comments. Please take a look.

File: core/trino-main/src/main/java/io/trino/security/AccessControlUtil.java (hunk 9 above)
Comment (kokosing) [reply core/trino-main/src/main/java/io/trino/security/AccessControlUtil.java:71]: Squashed.

File: testing/trino-tests/src/test/java/io/trino/server/testing/TestSystemSecurityMetadata.java [hunk 11]
```diff
@@ -11,7 +11,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package io.trino.security;
+package io.trino.server.testing;
```
Comment (findepi) [inline testing/trino-tests/src/test/java/io/trino/server/testing/TestSystemSecurityMetadata.java:14]: `testing` is for testing-related utilities that are (unfortunately) part of production jar

a test (TestSystemSecurityMetadata) should not be in the testing package, please move it back

File: core/trino-main/src/main/java/io/trino/security/AccessControlUtil.java [hunk 12]
```diff
@@ -43,19 +40,14 @@ public static List<BasicQueryInfo> filterQueries(Identity identity, List<BasicQu
...
+        Collection<Identity> allowedOwners = accessControl.filterQueriesOwnedBy(identity, owners);
 
-        Set<FullIdentityEquality> allowedOwners = owners.stream()
-                .map(FullIdentityEquality::new)
-                .collect(toImmutableSet());
         return queries.stream()
                 .filter(queryInfo -> {
                     Identity queryIdentity = queryInfo.getSession().toIdentity();
-                    return queryIdentity.getUser().equals(identity.getUser()) || allowedOwners.contains(new FullIdentityEquality(queryIdentity));
+                    return queryIdentity.getUser().equals(identity.getUser()) || allowedOwners.contains(queryIdentity);
```
Comment (lukasz-walkiewicz) [inline core/trino-main/src/main/java/io/trino/security/AccessControlUtil.java:50]: nit: previously it was a set whereas now it's just a collection so contains might be a bit slower (assuming the returned collection is not actually a set ofc)

**Review by lukasz-walkiewicz — APPROVED** [review]

**Comment by colebow** [comment]
Does this need a release note? @kokosing

**Comment by kokosing** [comment]
I don't think so.
